artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPMessage.java

```diff
@@ -606,22 +606,36 @@ public String getAddress() {
 
    @Override
    public AMQPMessage setAddress(String address) {
-      this.address = SimpleString.toSimpleString(address, coreMessageObjectPools == null ? null : coreMessageObjectPools.getAddressStringSimpleStringPool());
+      internalSetAddress(address);
+      setProtonAddress(address);
       return this;
    }
 
+   private void internalSetAddress(String address) {
+      this.address = SimpleString.toSimpleString(address, coreMessageObjectPools == null ? null : coreMessageObjectPools.getAddressStringSimpleStringPool());
+   }
+
    @Override
    public AMQPMessage setAddress(SimpleString address) {
       this.address = address;
+      setProtonAddress(address.toString());
```

Comment:
Where all is the outer setAddress method used? The AMQP Properties section is part of the immutable bare message so we shouldn't in general be setting the 'to' address in it or creating the section if they weren't present. Exception might be made during cases like protocol conversion, but it seems like it should be explicit rather than a side effect that might see unintended use as here.

Reply:
this is used on expiry or DLQ. you could argue those are copied messages and new messages. We actually copy the message during the DLQ or Expiry transfer.

I had the option to use extraProperties which only transverses through the broker... however there are implications on protocol conversions that could open further possibilities for bugs. this was the simpler change given it's copying the message.


Notice that setAddress will not be effective unless the message is reencoded.. so it only works on Expiry or DLQ.

Reply:
I added javadoc on Message.setAddress

Reply:
@gemmellr notice this has been the case for a while.. there was a bug on setAddress which was affecting customers.

Reply:
I must be misinterpretting things then. The setAddress methods (again, there are two) appear used already in several places of note within the broker.

For example, I interpret this bit as meaning every AMQP message received on a link with a target address will have it called on it, is that the case?
https://github.com/apache/activemq-artemis/blob/2.6.0/artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPSessionCallback.java#L448

If so, doesnt that mean that anyone reencoding the message for any reason, even if they have changed nothing, might find the payload changes due to the setAddress side effects?

Reply:
if the user calls reencode on a plugin, certainly the payload will be affected.

I'm changing setAddress to use extraproperties.. that will clear any confusion from this.

It wasn't the intend of this fix though.. but i will do it here.